source/patterns/@aws-solutions-constructs/core/lib/kendra-helper.ts

```diff
@@ -0,0 +1,254 @@
+ *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
+ *  and limitations under the License.
+ */
+
+/*
+ *  The functions found here in the core library are for internal use and can be changed
+ *  or removed outside of a major release. We recommend against calling them directly from client code.
+ */
+
+import * as kendra from 'aws-cdk-lib/aws-kendra';
+import * as iam from 'aws-cdk-lib/aws-iam';
+import { addCfnSuppressRules, consolidateProps } from "./utils";
+import { Aws } from 'aws-cdk-lib';
+
+// Note: To ensure CDKv2 compatibility, keep the import statement for Construct separate
+import { Construct } from 'constructs';
+import { DefaultKendraIndexProps } from './kendra-defaults';
+import { generatePhysicalName, overrideProps } from "./utils";
+
+export interface BuildKendraIndexProps {
+  readonly kendraIndexProps?: kendra.CfnIndexProps | any;
+  /**
+   * Existing instance of Kendra Index object, Providing both this and kendraIndexProps will cause an error.
+   *
+   * @default - None
+   */
+  readonly existingIndexObj?: kendra.CfnIndex;
+}
+
+/**
+ * @internal This is an internal core function and should not be called directly by Solutions Constructs clients.
+ */
+export function buildKendraIndex(scope: Construct, id: string, props: BuildKendraIndexProps): kendra.CfnIndex {
+  // Conditional lambda function creation
+  if (props.existingIndexObj) {
+    // The client provided an Index, so we'll do nothing and return it to them
+    return props.existingIndexObj;
+  } else {
+
+    let indexRoleArn: string = "";
+
+    // If the client provided a role, then don't bother creating a new one that we don't need
+    if (!props.kendraIndexProps?.roleArn) {
+      indexRoleArn = CreateKendraIndexLoggingRole(scope, id);
+    }
+    const defaultIndexProperties = DefaultKendraIndexProps(id, indexRoleArn);
+
+    const consolidatedIndexProperties = consolidateProps(defaultIndexProperties, props.kendraIndexProps);
+    const newIndex = new kendra.CfnIndex(scope, `kendra-index-${id}`, consolidatedIndexProperties);
+    addCfnSuppressRules(newIndex, [{
+      id: "W80",
+      reason: "We consulted the Kendra TFC and they confirmed the default encryption is sufficient for general use cases"
+    }]);
+
+    return newIndex;
+  }
+}
+
+/**
+ * @internal This is an internal core function and should not be called directly by Solutions Constructs clients.
+ */
+export function AddMultipleKendraDataSources(scope: Construct,
+  id: string,
+  kendraIndex: kendra.CfnIndex,
+  clientDataSourceProps: Array<Partial<kendra.CfnDataSourceProps>>): kendra.CfnDataSource[] {
+
+  const returnDataSources: kendra.CfnDataSource[] = [];
+  clientDataSourceProps.forEach((props, index) => {
+    returnDataSources.push(AddKendraDataSource(scope, `${id}${index}`, kendraIndex, props));
+  });
+  return returnDataSources;
```

Comment:
this could be refactored to something like - 
```
return clientDataSourceProps.map((props, index) => AddKendraDataSource(scope, `${id}${index}`, kendraIndex, props));
```